nexus/src/app/background/abandoned_vmm_reaper.rs

```diff
@@ -0,0 +1,450 @@
+    NonZeroU32::new_unchecked(100)
+};
+
+impl AbandonedVmmReaper {
+    pub fn new(datastore: Arc<DataStore>) -> Self {
+        Self { datastore }
+    }
+
+    /// List abandoned VMMs and clean up all of their database records.
+    async fn reap_all(
+        &mut self,
+        results: &mut ActivationResults,
+        opctx: &OpContext,
+    ) -> Result<(), anyhow::Error> {
+        slog::info!(opctx.log, "Abandoned VMM reaper running");
+
+        let mut paginator = Paginator::new(MAX_BATCH);
+        let mut last_err = Ok(());
+        while let Some(p) = paginator.next() {
+            let vmms = self
+                .datastore
+                .vmm_list_abandoned(opctx, &p.current_pagparams())
+                .await
+                .context("failed to list abandoned VMMs")?;
+            paginator = p.found_batch(&vmms, &|vmm| vmm.id);
+            self.reap_batch(results, &mut last_err, opctx, &vmms).await;
+        }
+
+        last_err
+    }
+
+    /// Clean up a batch of abandoned VMMs.
+    ///
+    /// This is separated out from `reap_all` to facilitate testing situations
+    /// where we race with another Nexus instance to delete an abandoned VMM. In
+    /// order to deterministically simulate such cases, we have to perform the
+    /// query to list abandoned VMMs, ensure that the VMM record is deleted, and
+    /// *then* perform the cleanup with the stale list of abandoned VMMs, rather
+    /// than doing it all in one go. Thus, this is factored out.
+    async fn reap_batch(
+        &mut self,
+        results: &mut ActivationResults,
+        last_err: &mut Result<(), anyhow::Error>,
+        opctx: &OpContext,
+        vmms: &[Vmm],
+    ) {
+        results.found += vmms.len();
+        slog::debug!(opctx.log, "Found abandoned VMMs"; "count" => vmms.len());
+
+        for vmm in vmms {
+            let vmm_id = vmm.id;
+            slog::trace!(opctx.log, "Deleting abandoned VMM"; "vmm" => %vmm_id);
+            // Attempt to remove the abandoned VMM's sled resource reservation.
+            match self.datastore.sled_reservation_delete(opctx, vmm_id).await {
+                Ok(_) => {
+                    slog::trace!(
+                        opctx.log,
+                        "Deleted abandoned VMM's sled reservation";
+                        "vmm" => %vmm_id,
+                    );
+                    results.sled_reservations_deleted += 1;
+                }
+                Err(e) => {
+                    slog::warn!(
+                        opctx.log,
+                        "Failed to delete sled reservation for abandoned VMM";
+                        "vmm" => %vmm_id,
+                        "error" => %e,
+                    );
+                    results.error_count += 1;
+                    *last_err = Err(e).with_context(|| {
+                        format!(
+                            "failed to delete sled reservation for {vmm_id}"
```

Comment:
I might add something to the error string that indicates the UUID refers to the VMM, something like `"...reservation for VMM {vmm_id}"`. Unless that context is already obvious from the existing error `e`...

Reply:
yeah, good call, thanks!